packages/grafana-data/src/field/scale.ts

```diff
@@ -1,127 +1,67 @@
-    // Calculate min/max if required
-    let min = config.min;
-    let max = config.max;
-
-    if (!isNumber(min) || !isNumber(max)) {
-      if (field.values && field.values.length) {
-        const stats = reduceField({ field, reducers: [ReducerID.min, ReducerID.max] });
-        if (!isNumber(min)) {
-          min = stats[ReducerID.min];
-        }
-        if (!isNumber(max)) {
-          max = stats[ReducerID.max];
-        }
-      } else {
-        min = 0;
-        max = 100;
-      }
-    }
-
-    const delta = max! - min!;
-
-    // Use a d3 color scale
-    let interpolator: colorInterpolator | undefined;
-
-    if (useColorScheme) {
-      interpolator = (d3 as any)[`interpolate${color!.schemeName}`] as colorInterpolator;
-    }
-
-    return (value: number) => {
-      const percent = (value - min!) / delta;
-      const threshold = thresholds
-        ? getActiveThreshold(percentThresholds ? percent * 100 : value, thresholds.steps)
-        : undefined; // 0-100
-      let color = fixedColor;
-
-      if (interpolator) {
-        color = interpolator(percent);
-      } else if (threshold) {
-        color = getColorFromHexRgbOrName(threshold!.color, themeType);
-      }
+  return (value: number) => {
+    const percent = (value - info.min!) / info.delta;
+    const threshold = getActiveThresholdForValue(field, value, percent);
 
-      return {
-        percent,
-        threshold,
-        color,
-      };
+    return {
+      percent,
+      threshold,
+      color: getColor(value, percent, threshold),
     };
-  }
+  };
+}
 
-  if (thresholds) {
-    return (value: number) => {
-      const threshold = getActiveThreshold(value, thresholds.steps);
-      const color = fixedColor ?? (threshold ? getColorFromHexRgbOrName(threshold.color, themeType) : undefined);
-      return {
-        threshold,
-        color,
-      };
-    };
-  }
+interface FieldMinMaxInfo {
+  min?: number | null;
+  max?: number | null;
+  delta: number;
+}
 
-  // Constant color
-  if (fixedColor) {
-    return (value: number) => {
-      return { color: fixedColor };
-    };
+function getMinMaxAndDelta(field: Field): FieldMinMaxInfo {
+  if (field.type !== FieldType.number) {
+    return { min: 0, max: 100, delta: 100 };
   }
 
-  // NO-OP
-  return (value: number) => {
-    return {};
-  };
-}
+  // Calculate min/max if required
+  let min = field.config.min;
+  let max = field.config.max;
 
-export function getActiveThreshold(value: number, thresholds: Threshold[]): Threshold {
-  let active = thresholds[0];
-  for (const threshold of thresholds) {
-    if (value >= threshold.value) {
-      active = threshold;
+  if (!isNumber(min) || !isNumber(max)) {
+    if (field.values && field.values.length) {
+      const stats = reduceField({ field, reducers: [ReducerID.min, ReducerID.max] });
+      if (!isNumber(min)) {
+        min = stats[ReducerID.min];
+      }
+      if (!isNumber(max)) {
+        max = stats[ReducerID.max];
+      }
```

Comment:
Personal preference of course, but I think it's a bit more readable with fewer nested ifs
```suggestion
      min = !isNumber(min) ? stats[ReducerID.min] : min;
      max = !isNumber(max) ? stats[ReducerID.max] : max;
```

Reply:
this was just old copied code, not written by me :)